### PR TITLE
[AMDGPU][GlobalIsel] Introduce isRegType to check for legal types, instead of checking bit width. 

### LINKED ISSUE #79578: GlobalISel: infinite loop trying to legalize G_BUILD_VECTOR
```
; RUN: llc -global-isel=1 -mtriple=amdgcn -mcpu=gfx1030 < %s
define void @main(<19 x i32> %arg) {
bb:
  %i = bitcast <19 x i32> %arg to <38 x i16>
  %i1 = extractelement <38 x i16> %i, i64 0
  %i2 = icmp eq i16 %i1, 0
  %i3 = zext i1 %i2 to i32
  %i4 = bitcast i32 %i3 to float
  %i5 = insertelement <4 x float> zeroinitializer, float %i4, i64 0
  call void @llvm.amdgcn.image.store.2d.v4f32.i32(<4 x float> %i5, i32 0, i32 0, i32 0, <8 x i32> zeroinitializer, i32 0, i32 0)
  ret void
}
declare void @llvm.amdgcn.image.store.2d.v4f32.i32(<4 x float>, i32 immarg, i32, i32, <8 x i32>, i32 immarg, i32 immarg)
```

### PR DESCRIPTION
**Pull request by sstipano**
In D151116 it was suggested to have a set of classes to cover every possible case. This does it for bitcast first.

closes #79578

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 1]
```diff
@@ -272,6 +272,104 @@ static LegalityPredicate elementTypeIsLegal(unsigned TypeIdx) {
...
+
+  std::initializer_list<LLT> AllPtrTypes{V2FlatPtr,   V3LocalPtr,  V5LocalPtr,
+                                         V16LocalPtr, V2GlobalPtr, V4GlobalPtr};
+
+  return typeInSet(Ty, AllS32Vectors) || typeInSet(Ty, AllS64Vectors) ||
+         typeInSet(Ty, AllScalarTypes) || typeInSet(Ty, AllS16Vectors) ||
+         typeInSet(Ty, AllPtrTypes) || Ty.isPointer();
+}
+
+static LegalityPredicate isRegType(unsigned TypeIdx,
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:366]: Needs a comment explaining what this is. In particular how is it different from isRegisterType and why do we need both (and can we have better names for them)?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/legalize-build-vector.mir [hunk 2]
```diff
@@ -299,129 +299,6 @@ body: |
     S_NOP 0, implicit %12
 ...
 ---
-name: legal_v13s32
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/legalize-build-vector.mir:302]: Why are these removed?

File: llvm/test/CodeGen/AMDGPU/GlobalISel/extractelement.ll [hunk 3]
```diff
@@ -2714,6 +2714,168 @@ entry:
   ret double %ext
 }
 
+define amdgpu_ps double @dyn_extract_v7f64_s_v_bitcast(<14 x float> inreg %userData, i32 %sel) {
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/extractelement.ll:2717]: Can you precommit these tests, or do they fail to compile without your patch?

**Comment by jayfoad** [comment]
There is also some relevant discussion on https://reviews.llvm.org/D148096 which introduced isIllegalRegisterType.

Do we really need both isIllegalRegisterType and isRegType?

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp (hunk 1 above)
Comment (sstipanovic) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:366]: isRegisterType check is not sufficient. For example it considers v13s32 as legal which doesn't have corresponding reg class. If this approach is accepted, this would be the first step to replace isRegisterType with isRegType (we can name it differently, of course). The idea is to replace it step by step because it is not as straight forward as replacing all calls to isRegisterType with isRegType.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/extractelement.ll (hunk 3 above)
Comment (sstipanovic) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/extractelement.ll:2717]: They fail without this patch

File: llvm/test/CodeGen/AMDGPU/GlobalISel/legalize-build-vector.mir (hunk 2 above)
Comment (sstipanovic) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/legalize-build-vector.mir:302]: Because these are not legal anymore

**Comment by sstipanovic** [comment]
> There is also some relevant discussion on https://reviews.llvm.org/D148096 which introduced isIllegalRegisterType.
> 
> Do we really need both isIllegalRegisterType and isRegType?

I think once we replace all usage of isRegisterType we could do without isIllegalRegisterType.

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp (hunk 1 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:366]: That sounds like a good plan. Could you add a TODO comment to isRegisterType saying that all uses should be migrated to isRegType?

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 4]
```diff
@@ -272,6 +272,104 @@ static LegalityPredicate elementTypeIsLegal(unsigned TypeIdx) {
...
+
+  // TODO: list all possible ptr vectors
+  const LLT V2FlatPtr = LLT::fixed_vector(2, FlatPtr);
+  const LLT V3LocalPtr = LLT::fixed_vector(3, LocalPtr);
+  const LLT V5LocalPtr = LLT::fixed_vector(5, LocalPtr);
+  const LLT V16LocalPtr = LLT::fixed_vector(16, LocalPtr);
+  const LLT V2GlobalPtr = LLT::fixed_vector(2, GlobalPtr);
+  const LLT V4GlobalPtr = LLT::fixed_vector(4, GlobalPtr);
+
+  std::initializer_list<LLT> AllPtrTypes{V2FlatPtr,   V3LocalPtr,  V5LocalPtr,
```
Comment (petar-avramovic) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:358]: What should we do with vectors of pointers, list all possible vector sizes that fit in some register class?

**Comment by sstipanovic** [comment]
@jayfoad ping

File: llvm/test/CodeGen/AMDGPU/GlobalISel/legalize-build-vector.mir (hunk 2 above)
Comment (jayfoad) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/legalize-build-vector.mir:302]: But this is a legalizer test. The legalizer is supposed to legalize them.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/legalize-build-vector.mir [hunk 5]
```diff
@@ -299,129 +299,6 @@ body: |
...
-    %5:_(s32) = COPY $vgpr5
-    %6:_(s32) = COPY $vgpr6
-    %7:_(s32) = COPY $vgpr7
-    %8:_(s32) = COPY $vgpr8
-    %9:_(s32) = COPY $vgpr9
-    %10:_(s32) = COPY $vgpr10
-    %11:_(s32) = COPY $vgpr11
-    %12:_(s32) = COPY $vgpr12
-    %13:_(<13 x s32>) = G_BUILD_VECTOR %0, %1, %2, %3, %4, %5, %6, %7, %8, %9, %10, %11, %12
-    S_NOP 0, implicit %13
```
Comment (petar-avramovic) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/legalize-build-vector.mir:338]: There is nothing we can do here, user of %13:_(<13 x s32>) is not legalized. There is nothing that could combine away %13:_(<13 x s32>)
Test was meant to show that G_BUILD_VECTOR is legal for <13 x s32> which is no longer the case

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/legalize-build-vector.mir:338]: Yes, this would need a legalizable user. We should teach the verifier to reject unlegalizable uses where there's no possible register class

**Comment by sstipanovic** [comment]
@jayfoad ping

**Comment by jayfoad** [comment]
> > There is also some relevant discussion on https://reviews.llvm.org/D148096 which introduced isIllegalRegisterType.
> > Do we really need both isIllegalRegisterType and isRegType?
> 
> I think once we replace all usage of isRegisterType we could do without isIllegalRegisterType.

Could you add a TODO comment to isIllegalRegisterType please?

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp (hunk 4 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:364]: I guess so. Can you write test cases for legalizing bitcast, that fail because you don't handle all pointer vector types here yet?

**Review by jayfoad — COMMENTED** [review]
I've just taken a fresh look at this. I do have a couple of concerns:

1. It introduces yet another list of supported register classes, which will need to be updated whenever we add a new one. But maybe that is unavoidable.
2. It looks like `isRegisterClassType` will not be super fast, especially since it has to construct all the pointer types on every call.

**Comment by sstipanovic** [comment]
> 2. It looks like `isRegisterClassType` will not be super fast, especially since it has to construct all the pointer types on every call.
@jayfoad  `initializer_list` makes things tricky. I can make pointer types static to avoid having them constructed on each call. What do you think?

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 6]
```diff
@@ -276,6 +277,105 @@ static LegalityPredicate elementTypeIsLegal(unsigned TypeIdx) {
...
+  return llvm::is_contained(Types, Ty);
+}
+
+static LLT GetAddrSpacePtr(unsigned AS, const GCNTargetMachine &TM) {
+  return LLT::pointer(AS, TM.getPointerSizeInBits(AS));
+}
+
+// Checks whether a type is in the list of legal register types.
+static bool isRegisterClassType(LLT Ty, const GCNTargetMachine &TM) {
+  const LLT GlobalPtr = GetAddrSpacePtr(AMDGPUAS::GLOBAL_ADDRESS, TM);
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:352]: How about, as the first thing in this function, if Ty is a pointer or vector-of-pointers, convert it to the equivalent non-pointer type? Would that work? Then we would not need to construct all these pointer types here.
```
  Ty = Ty.changeElementType(LLT::Scalar(Ty.getScalarSizeInBits()));
```

Comment (sstipanovic) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp]: This was a very good idea, thanks! I think it is much cleaner now.
Should we still have tests for different address spaces? Because now every ptr and ptr vector type is handled.

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 7]
```diff
@@ -276,6 +279,95 @@ static LegalityPredicate elementTypeIsLegal(unsigned TypeIdx) {
...
+  return llvm::is_contained(Types, Ty);
+}
+
+static LLT GetAddrSpacePtr(unsigned AS, const GCNTargetMachine &TM) {
+  return LLT::pointer(AS, TM.getPointerSizeInBits(AS));
+}
+
+// Checks whether a type is in the list of legal register types.
+static bool isRegisterClassType(LLT Ty) {
+  if (Ty.isVector() && Ty.getElementType().isPointer())
```
[the remark this answers is not in the document]
Comment (sstipanovic) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:349]: Which one should go in first?

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 8]
```diff
@@ -276,6 +279,89 @@ static LegalityPredicate elementTypeIsLegal(unsigned TypeIdx) {
...
+
+static std::initializer_list<LLT> AllS32Vectors = {
+    V2S32, V3S32,  V4S32,  V5S32,  V6S32,  V7S32, V8S32,
+    V9S32, V10S32, V11S32, V12S32, V16S32, V32S32};
+
+static std::initializer_list<LLT> AllS64Vectors = {V2S64, V3S64, V4S64, V5S64,
+                                                   V6S64, V7S64, V8S64, V16S64};
+
+static bool typeInSet(LLT Ty, std::initializer_list<LLT> Types) {
+  return llvm::is_contained(Types, Ty);
```
[the remark this answers is not in the document]
Comment (sstipanovic) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp]: Agreed, changed everything to is_contained.

**Review by jayfoad — APPROVED** [review]
Seems reasonable to me, thanks.
